### ISSUE TITLE
Add tag configuration to track custom dimensions

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#7600]
## Feature Description

Necessary tag configuration should be added in the `analytics-4` module so that the available custom dimensions for the current GA property are tracked.

The custom dimensions can be tracked by passing an object containing them to the `gtag("config", "G-XXXX")` script, for example:

```js
gtag(
	"config",
	"G-XXXX",
	{
		"googlesitekit_post_type": "page",
		"googlesitekit_post_author": "1",
		"googlesitekit_post_categories": "1,2,3",
		"googlesitekit_post_date": "20220614",
	}
);
```

See relevant section in the design doc](https://docs.google.com/document/d/1htqStvvyXq_eg5GsHSbdBBgm4Tu_Z6KXsIOBfjWQ5oM/edit?resourcekey=0-Px-yDiCQOTz30DBKllhU3g#heading=h.8rpbnsc62d64).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- If the `newsKeyMetrics` feature flag is enabled and the connected GA4 property has available custom dimensions:
  - The tag configuration of the `analytics-4` module should be updated for both web & AMP to track data for the available custom dimensions.
  - The custom dimensions should only be tracked for singular (i.e. [is_singular](https://developer.wordpress.org/reference/functions/is_singular/)) templates, i.e. a single page/post should be tracked, but not a blog index or an archive.
  - The `googlesitekit_post_date`, `googlesitekit_post_author`, and `googlesitekit_post_categories` custom dimensions should only be tracked for posts with the `post` post type. This post type should be filterable (using `apply_filters` for example) so that the post type(s) that track these custom dimensions would be flexible.
  - The value for each possible custom dimension (for now) is described below:
    - For `googlesitekit_post_date`: The date, in `YYYYMMDD` format (i.e. what Analytics uses for its date dimension), on which the current WordPress post was published.
    - For `googlesitekit_post_author`: The user ID of the author of the current post, as a string.
    - For `googlesitekit_post_categories`: The categories of the current post as a comma-separated string of IDs of categories.
    - For `googlesitekit_post_type`: Post type of the current post - `post`, `page`, `attachment`, etc.

## Implementation Brief

- In `includes/Modules/Analytics_4.php`, create a new `get_custom_dimensions_data()` method which should do the following:
    * Use the Analytics_4 Settings class (get method) to check the value of the `availableCustomDimensions` for the connected GA property.
    * If there are no custom dimensions within the above setting, return an empty array. 
    * Otherwise, create and apply a new `custom_dimension_valid_post_types` filter with a default value of "post". Call the [`get_post_type()`](https://developer.wordpress.org/reference/functions/get_post_type/) WordPress function and check if this value is within the filtered array. If not, return an empty array.
    * Finally, call the `is_singular()` WordPress function. If the value is false, return an empty array.
    * Otherwise, for each of the custom dimensions within the above setting, build an array as follows:
      - Each available custom dimension name should be added to the array as keys.
      - The value for each possible custom dimension key should be added as follows:
        - For `googlesitekit_post_date`: The date on which the current WordPress post was published, obtainable using the [`get_the_date('Ymd')`](https://developer.wordpress.org/reference/functions/get_the_date/) WordPress function. This should yield a date in the format similar to what Analytics uses for its date dimension, e.g. `20220614`.
        - For `googlesitekit_post_author`: The user ID of the author of the current post, obtainable using the [`get_the_author_meta('ID')`](https://developer.wordpress.org/reference/functions/get_the_author_meta/) WordPress function. The user ID is a non-zero integer value.
        - For `googlesitekit_post_categories`: The categories of the current post, obtainable using the [`get_the_category()`](https://developer.wordpress.org/reference/functions/get_the_category/) WordPress function, which returns an array of [`WP_Term`](https://developer.wordpress.org/reference/classes/wp_term/) objects. The category ID can be accessed by using the `cat_ID` property on each object. This custom dimension value should be a comma-separated list of IDs of categories assigned to a post, e.g. `1, 3, 4, 5`.
        - For `googlesitekit_post_type`: Raw type of the current post, obtainable using the [`get_post_type()`](https://developer.wordpress.org/reference/functions/get_post_type/) WordPress function. Example, `post`, `page`, `attachment`, etc.
    * Return this array of dimension keys and values.
- In `includes/Modules/Analytics_4/AMP_Tag.php`:
  * Add a new private property, `custom_dimensions`. 
  * Create a public setter method, called `set_custom_dimensions()` which simply takes in an associative array of custom dimensions and sets it to the above property.
  * Override the `get_tag_config()` method and ensure the shape should look like follows, as also mentioned in the [design doc](https://docs.google.com/document/d/1htqStvvyXq_eg5GsHSbdBBgm4Tu_Z6KXsIOBfjWQ5oM/edit?resourcekey=0-Px-yDiCQOTz30DBKllhU3g#heading=h.8rpbnsc62d64):
    ```php
    array(
      $this->tag_id => array(
        'groups' => 'default',
        'linker' => array(
          'domains' => array( $this->home_domain ),
        ),
        'googlesitekit_post_author' => 1,
        'googlesitekit_post_categories' => "1,2,3",
        'googlesitekit_post_date' => "20220614",
        'googlesitekit_post_type' => "post"
      ),
    );
    ``` 
- In `includes/Modules/Analytics_4/Web_Tag.php`:
  * Add a new private property, `custom_dimensions`. 
  * Create a public setter method, called `set_custom_dimensions()` which simply takes in an associative array of custom dimensions and sets it to the above property.
  * Similar to the pattern followed within `Analytics/AMP_Tag` and `Analytics_4/AMP_Tag`, extract out the fetching/building of `gtag_opt` out of the `Analytics/Web_Tag::enqueue_gtag_script()` method and override it in `Analytics_4/Web_Tag`. Ensure this method is called consistently when both UA and GA4 are connected, or if only GA4 is connected. (Currently, if UA is enabled, then we do not add config options and do not call the filter correctly within `Analytics_4/Web_Tag::enqueue_gtag_script()`).
- In `includes/Modules/Analytics_4.php`, within the `register_tag()` method, call `get_custom_dimensions_data()` and inside the `can_register()` if block, call the `set_custom_dimensions()` method on `$tag` and pass in the custom dimensions that have been fetched.

### Test Coverage

* Add tests for the `get_custom_dimensions_data()` method. 

## QA Brief

* Follow QAB from #7598 to get custom dimensions configured on your site
* Go to a single post page on your site and verify that custom dimensions added to the gtag config as shown in this ticket description.
* Go to another page that is not a single post and ensure that there is no custom dimensions for that page in the gtag config.
* Also verify that the correct dimensions are passed in the AMP mode and when both UA and GA4 modules are connected.

## Changelog entry

* Add Analytics tag configuration to track custom dimensions for posts.

**Comment by techanvil** [comment]
Hey @nfmohit, this AC is basically a solid IB. Could you please redo the AC to describe what we're trying to achieve here in higher level language (it doesn't need to go into too much detail), and move the current AC to the IB section?

**Comment by nfmohit** [comment]
> Hey @nfmohit, this AC is basically a solid IB. Could you please redo the AC to describe what we're trying to achieve here in higher level language (it doesn't need to go into too much detail), and move the current AC to the IB section?

Thank you for pointing that out, @techanvil. I've updated the ACs to describe our end-goal in a higher level language. Please let me know if this looks good now.

**Comment by techanvil** [comment]
Thanks @nfmohit. Looks good, one thing I would suggest is that the last top-level point and its subpoints are duplicated in the IB and so I'd suggest they should be simplified in the AC along these lines (seeing as the detail is captured in the draft IB):

  - The value for each possible custom dimension (for now) is described below:
    - For `googlesitekit_post_date_created`: The date, in `YYYYMMDD` format, on which the current WordPress post was created.
    - For `googlesitekit_post_author`: The user ID of the author of the current post, as a string.
    - For `googlesitekit_post_categories`: The categories of the current post as a comma-separated string of IDs of categories.
    - For `googlesitekit_post_type`: Post type of the current post - `post`, `page`, `attachment`, etc.

**Comment by techanvil** [comment]
Thanks @nfmohit. AC LGTM :white_check_mark:

**Comment by eugene-manuilov** [comment]
Thanks for IB, @jimmymadon, but this is not how it should work. What we need to do here is to update `AMP_Tag` and `Web_Tag` classes in the `Google\Site_Kit\Modules\Analytics_4` namespace to have an additional `set_available_custom_dimensions` method that assigns associated array of custom dimensions and their values to a private property and then those classes should use that property to add it to configs if that is not empty.

The reason why we should do it that way is that that is not those classes responsibility to read and prepare data, those classes are responsible for rendering tags based on provided information. Thus we need to prepare and provide that information in another place that is responsible for usage of those classes. In our case, that is `Analytics_4::register_tag` who uses them which means that we need to add a new method to the `Analytics_4` class itself to prepare data for custom dimensions. That won't be perfect, probably would be better to have a class that "orchestrate" tags for Analytics 4 module, but we have what we have and currently we need to extend that `Analytics_4` class.

Basically, what you need to change in your IB is to replace that trait with a new method in the `Analytics_4` class, explain how to use that new method and what changes should be done in tag classes.

**Comment by eugene-manuilov** [comment]
> However, for Analytics_4/AMP_Tag, we could override the get_tag_config() method in to merge the custom dimensions.

@jimmymadon, yes, this sounds like a better approach to me than adding a new hook.

> And for `Analytics_4/Web_Tag`, we could extract out the fetching/building of `gtag_opt` out of the `Analytics/Web_Tag::enqueue_gtag_script()` method and override it in the child class like I mentioned for AMP above.

Yep, i like this. Plus, we need to make config rendering in the child class (for `Analytics_4`) to be consistent with the way how we render config in the parent class.

**Comment by eugene-manuilov** [comment]
Thanks, @jimmymadon. IB looks great! 👍

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@eugene - there seems to be a difference between the AC and QAB. 

AC: `The custom dimensions should only be tracked for singular templates, i.e., a single page/post should be tracked, but not a blog index or an archive.`   
QAB: `Go to another page that is not a single post and ensure that there is no custom dimensions for that page in the gtag config.`

Here's what I see on a page.  

The tag for `googlesitekit_post_type":"page"` appears.   From the QAB, this suggests that this shouldn't appear. 

A single page also has an author and publish date, so I wondered if we should see all the appropriate tags here or nothing. 

![image](https://github.com/google/site-kit-wp/assets/73545194/b2c70fee-8458-4f4d-a54d-9b65f23f6e86)

Everything looks fine on a post; no gtags appear on an archive page.

**Comment by eugene-manuilov** [comment]
@wpdarren, hm.. Why do you have `googlesitekit_post_type` in the config? 🤔 I can't reproduce it on my end. Do you have a custom hook for the `custom_dimension_valid_post_types` filter?

**Comment by techanvil** [comment]
@eugene-manuilov @wpdarren actually, reviewing the AC vs the IB (and [implementation](https://github.com/google/site-kit-wp/pull/7694/files#diff-057a7c0b7d3ca6ea365f58acd6913b97a0b3bd5fa2bd139b1bb492fd4e7d1725R1292-R1308)) on this one, it looks like there's a bit of a mismatch. According to the AC, the `googlesitekit_post_date`, `googlesitekit_post_author`, and `googlesitekit_post_categories` should not be tracked when the current page is a singular `post` type, the implication there being that `googlesitekit_post_type` should still be tracked regardless.

However the IB specifies to avoid tracking _all_ of the custom dimensions when the page is not a single `post`.

It looks like we might need to revisit this one to ensure that `googlesitekit_post_type` is tracked in all cases.

@nfmohit, would you mind confirming this one?

**Comment by nfmohit** [comment]
Hi folks.

1. The custom dimensions should only be tracked if the current page is a "singular" template, i.e. `is_singular()` returns true.
2. The `googlesitekit_post_date`, `googlesitekit_post_author`, and `googlesitekit_post_categories` custom dimensions should only be tracked if the current singular template is of a `post` post type. The `googlesitekit_post_type` custom dimension should be tracked for singular templates of all post types.

Please let me know if that clarifies the aim. Thank you!

**Comment by techanvil** [comment]
Thanks for clarifying this, @nfmohit. Despite @wpdarren seeing a `page` being tracked, this is contrary to the expectations of the current code. I suspect this needs a followup PR to ensure the logic matches AC. @eugene-manuilov, please can you take a look?

**Comment by eugene-manuilov** [comment]
@techanvil, can you review #7713?

**Comment by techanvil** [comment]
Back with you for another pass, @wpdarren.

**Comment by aaemnnosttv** [comment]
@techanvil @eugene-manuilov please update this filter to have our `googlesitekit` prefix, thanks!
https://github.com/google/site-kit-wp/blob/7be98e574f3bfce9cd041c5798b8239544d44b1a/includes/Modules/Analytics_4.php#L1382-L1389

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, good spot.

@eugene-manuilov, as you did the execution for this would you like to submit a followup PR to address this and Evan's additional [comment](https://github.com/google/site-kit-wp/pull/7694/files#r1371080954), and I can then review?

@wpdarren, I will move this back to execution in the meantime and send back to you in QA when it's updated.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@techanvil @aaemnnosttv when a site has UA and GA4 connected what should I expect to see for a post and page related to the custom dimensions? 

This is what I see for a post
![image](https://github.com/google/site-kit-wp/assets/73545194/434c3df1-4a40-49ee-8fc4-9414efd1e6fd)

This is what I see for a page
![image](https://github.com/google/site-kit-wp/assets/73545194/058ff5d9-c264-4716-8be3-d1508823ebe7)

I know Evan mentioned in a stand-up that he felt this wasn't correct, so wanted to check. 

**Verified:**

- On a single post page on your site the custom dimensions added to the gtag config as shown in this ticket description.
- On a page that is not a single post there is no custom dimensions for that page in the gtag config.
- The correct dimensions are passed in the AMP mode snippet. See screenshot.


<details>
 <summary>Screenshots</summary>

![image](https://github.com/google/site-kit-wp/assets/73545194/679474ff-b90b-4cef-8c06-8938efeb0d3f)
![image](https://github.com/google/site-kit-wp/assets/73545194/78a2108f-bf9b-4ebd-b1f3-5e90d02363fc)
![image](https://github.com/google/site-kit-wp/assets/73545194/e392dbc5-401d-4485-9a99-0ed2f4355b71)

</details>

**Comment by techanvil** [comment]
> ### QA Update: ⚠️
> @techanvil @aaemnnosttv when a site has UA and GA4 connected what should I expect to see for a post and page related to the custom dimensions?
> 
> This is what I see for a post ![image](https://user-images.githubusercontent.com/73545194/278328305-434c3df1-4a40-49ee-8fc4-9414efd1e6fd.png)
> 
> This is what I see for a page ![image](https://user-images.githubusercontent.com/73545194/278328575-058ff5d9-c264-4716-8be3-d1508823ebe7.png)
> 
> I know Evan mentioned in a stand-up that he felt this wasn't correct, so wanted to check.

Hi @wpdarren, that does look OK to me - we're outputting the [`config`](https://developers.google.com/tag-platform/gtagjs/reference#config) for each of the properties, but only adding the custom dimensions to the config for the GA4 property.

Furthermore the choice of custom dimensions is as per the AC, with `googlesitekit_post_type` being tracked in both cases (for the singular `post` and `page` respectively), but `googlesitekit_post_author`, `googlesitekit_post_date` and `googlesitekit_post_categories` only being tracked for the `post`.

I'm not sure if I might have missed something that @aaemnnosttv has spotted but from my understanding it does seem to be alright.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil, this looks correct to me as well. The issue I flagged to check here was regarding an observation I had with our data plugin before this was added to SK, where the parameters were only being added to the UA tag when both were present, which was a problem. Now we're only seeing them added to the GA4 tag which is fine, since we're only reporting on these with GA4, not to mention UA tagging is being removed entirely soon and probably isn't logging hits for anyone anymore anyways. I also confirmed that the parameters go through as expected when the `config` is using a `GT-` Google Tag ID rather than the GA4 measurement ID 👍

**Comment by wpdarren** [comment]
### QA Update: ✅ 

Thank you @techanvil @aaemnnosttv for the clarification. 

**Verified:**

- On a single post page on your site the custom dimensions added to the gtag config as shown in this ticket description.
- On a page that is not a single post there is no custom dimensions for that page in the gtag config.
- The correct dimensions are passed in the AMP mode snippet. See screenshot.
- The custom dimensions are being added to the GA4 only when a UA and GA4 property are connected.

<details>
 <summary>Screenshots</summary>

![image](https://github.com/google/site-kit-wp/assets/73545194/8b8e0086-500d-4657-afea-0a0b68a5cb5a)
![image](https://github.com/google/site-kit-wp/assets/73545194/64fa0706-cca9-450b-bf75-723b5474ed71)
![image](https://github.com/google/site-kit-wp/assets/73545194/679474ff-b90b-4cef-8c06-8938efeb0d3f)
![image](https://github.com/google/site-kit-wp/assets/73545194/78a2108f-bf9b-4ebd-b1f3-5e90d02363fc)
![image](https://github.com/google/site-kit-wp/assets/73545194/e392dbc5-401d-4485-9a99-0ed2f4355b71)

</details>
